PRO-938 Intro schema (#53)
apostrophecms/a3-docs: docs/guide/introduction.md

::: note
You will see a `trash` field in the log output. The interface does not show this as a normal field, but it is registered as one to support editing via the REST API.
:::

#### Reusing an inherited group's name will ungroup its original fields

For example, `title` is in the default "Basics" group. If you add a `basics` group in your field configuration and do not include `title` in its `fields` array, it will no longer be in any group. Fields that are not part of any group will appear in an "Ungrouped" tab in the interface.

As in the example above, you could include `title` with the "Basics" group along with new fields.

```js
// modules/product/index.js
module.exports = {
  // ...
  fields: {
    add: {
      // ...
    },
    group: {
      basics: {
        label: 'Basics',
        fields: [ 'title', 'price', 'description', 'image' ]
      }
    }
  }
};
```

#### You don't need a `basics` group

There is nothing special about "Basics." It is a default group name, but if you place all of the fields from that group (or any inherited group) in a new one the group will no longer appear in the UI.

> `utility` _is_ a special group. It places fields in the right column of the content editor interface. You are allowed to add fields to that group and move existing fields.

Author: I went in this order since the example involves Basics. Utility hasn't been mentioned to that point other than in a code block comment.

## Content schemas

> The content field schema is another key concept that most Apostrophe development will use. The "schema" here simply refers to definition of fields allowed in a particular context (e.g., a page, piece, or widget type). This is important so that the user interface knows what fields to show editors. It also tells both the UI and server how to validate data for that context.

Reviewer: ```suggestion
The content field schema is another key concept in Apostrophe. The "schema" simply refers the content fields that define a particular page, content type, or widget.```

Author: Until we're using "content type" universally we should use "piece type" for consistency.

> Our blog post module might look something like this with its field schema:

Reviewer: ```suggestion
The field schema of a blog post might look something like this:
```

Author: "Our blog post module" refers to the example in the previous section. Also is to be a bit less dry.

```javascript
// modules/blog-post/index.js
module.exports = {
  extend: '@apostrophecms/piece-type',
  fields: {
    add: {
      authorName: {
        label: 'Author name',
        type: 'string'
      },
      body: {
        label: 'Blog post body',
        type: 'area',
        options: {
          widgets: {
            '@apostrophecms/rich-text': {}
          }
        }
      }
    },
    group: {
      blogFields: [ 'authorName', 'body' ]
    }
  }
};
```

> The `fields` setting is the parent property for the schema configuration. In this case, it has two subsections: `add`, where fields are added to the schema, and `group`, which organizes the fields for the user interface.

Reviewer: ```suggestion
All schemas are configured within the `fields` setting. In this case, `fields` has two subsections: `add`, where fields are added to the schema, and `group`, which organizes the fields for the user interface. All schemas have some required fields, such as `title` and `slug`, which are added and grouped by default.
```

Author: The last bit isn't true for widgets.

> Each property in the `add` object is a field you are including in the schema. Each property in `group` is a section of the interface, set to an array of fields to include in that section. In both cases, there are default fields, such as `title`, that are added and grouped for you already.

Reviewer: ```suggestion
```

I suggest removing this section, and adding the last sentence to the first paragraph.

Author: If we remove this we aren't telling them what the subsections are doing.

> There are other elements to schemas, but that is generally how developers define them. There is additional reference documentation on [the `fields` setting](/reference/module-api/module-overview.md#fields) and [individual field types](/reference/field-types/).

Reviewer: ```suggestion
For additional reference documentation on [the `fields` setting](/reference/module-api/module-overview.md#fields) and [individual field types](/reference/field-types/).
```

Reviewer: We don't need "there are other things" if we're leading them to "additonal"

## Content schemas

The content field schema is another key concept in Apostrophe. The "schema" simply refers to the content fields for a particular page, piece, or widget type. In addition to setting fields for the user interface, it supports data validation.

Our blog post module's schema might look something like this:

```javascript
// modules/blog-post/index.js
module.exports = {
  extend: '@apostrophecms/piece-type',
  fields: {
    add: {
      authorName: {
        label: 'Author name',
        type: 'string'
      },
      body: {
        label: 'Blog post body',
        type: 'area',
        options: {
          widgets: {
            '@apostrophecms/rich-text': {}
          }
        }
      }
    },
    group: {
      blogFields: [ 'authorName', 'body' ]
    }
  }
};
```

Content schemas are configured in the `fields` setting. In this case, `fields` has two subsections: `add`, where fields are added to the schema, and `group`, which organizes the fields for the user interface.

> Each property in the `add` object is a field you are including in the schema. Each property in `group` is a section of the interface, set to an array of fields to include in that section.

Author: You'd suggested this before. I noted that if we remove this we aren't telling them what is happening in the example. The previous paragraph is only the most basic description of their purpose.

See the reference documentation on [the `fields` setting](/reference/module-api/module-overview.md#fields) and [individual field types](/reference/field-types/) for more information.

### Using existing field groups

> Fields that a piece type inherits will likely already be in field groups. This is certainly true for the default fields, `title`, `slug`, and `visibility`. You can add new fields into these groups and rearrange them if needed. There are a few things to keep in mind as you do.

Author: I'm trying to keep sentences short to be easier to read. I'll rework this, though.

> **It's fairly simple to see what the existing groups are.** Working with inherited fields and field groups is harder when you don't know what they are. You can log them in your terminal easily from the new module's [initialization function](/reference/module-api/module-overview.md#initialization-function).

Reviewer: @abea I thought we wanted to avoid terms like "easily". Additionally, pointing out that its "hard" adds fluff between "its fairly simple" and "you can log them". If you want to keep the sentence about difficulty, it might be better do try this

```suggestion
Working with inherited fields and field groups can be difficult when you don't know what they are. Fortunately, it's fairly simple to identify existing groups by logging them in your terminal from the new module's [initialization function](/reference/module-api/module-overview.md#initialization-function).
```

> **It's fairly simple to see what the existing groups are.** Working with inherited fields and field groups is harder when you don't know what they are. You can log them in your terminal easily from the new module's [initialization function](/reference/module-api/module-overview.md#initialization-function).

Author: I'm using a structure of the leading bold sentences, but I'll change them to H4s to be clearer.

Good point about "easily." I hedged with "fairly simple," but we should remove the other.

```javascript
// modules/product/index.js
module.exports = {
  extend: '@apostrophecms/piece-type',
  init (self) {
    console.log(self.fieldsGroups);

    // Output:
    // {
    //   basics: { label: 'Basics', fields: [ 'title' ] },
    //   utility: { fields: [ 'slug' ] },
    //   permissions: { label: 'Permissions', fields: [ 'visibility' ], last: true }
    // }
  }
};
```

The `init` function runs once on start up and has access to the module as an argument. By the time it runs, the field groups have been compiled into an object named `fieldsGroups`. If you haven't added any fields yet you can log this to see what you are working with.

::: note
You will see a `trash` field in the log output. The interface does not show this as a normal field, but
:::

> **If you name an existing field group in your configuration, the fields in that group will be ungrouped.** For example, `title` is in the default "Basics" group. If you add a `basics` group in your field configuration and do not include `title` in its `fields` array, it will no longer be in any group. Fields that are not part of any group will appear in an "Ungrouped" tab in the interface.

Author: Yes, that's correct. It's annoying. I wish there was a better way.

> **If you name an existing field group in your configuration, the fields in that group will be ungrouped.** For example, `title` is in the default "Basics" group. If you add a `basics` group in your field configuration and do not include `title` in its `fields` array, it will no longer be in any group. Fields that are not part of any group will appear in an "Ungrouped" tab in the interface.

Reviewer: @abea Yeah.  I definitely found myself thinking "why is it this way" as I was reading. For now though:

```suggestion
**Using an existing group name in your configuration will cause fields already in that group to become "ungrouped"**. For example, `title` is in the default "Basics" group. If you add a `basics` group in your field configuration and do not include `title` in its `fields` array, it will no longer be in any group. Fields that are not part of any group will appear in an "Ungrouped" tab in the interface.
```
